src/levanter/models/hyena.py

```diff
@@ -0,0 +1,540 @@
+"""An implementation of the Hyena operator.
+
+Paper: [Hyena Hierarchy: Towards Larger Convolutional Language Models](https://arxiv.org/abs/2302.10866)
+Official implementation in PyTorch:
+https://github.com/HazyResearch/safari/blob/541902aca88cb11af4d816ac762f3303e4ff8eea/src/models/sequence/hyena.py
+
+Current diffences from the official impl:
+- We don't support inner_factor.
+- We don't support post_order_ffn.
+- We don't support num_heads (the PyTorch impl's support for multiple heads seems incomplete, or at least
```

Comment:
I could be wrong about this. LMK if you can tell if their code actually supports multiple heads, and if so please explain it to me and I will try to port it over.

Reply:
doesn't it support num_heads:

https://github.com/HazyResearch/safari/blob/541902aca88cb11af4d816ac762f3303e4ff8eea/src/models/sequence/hyena.py#L317-L321

Reply:
oh right you said you don't really speak einsum. 

In their code:

```
rearrange(uc, 'b (ho v) (z l) -> b ho v z l', 
            z=self.num_blocks, 
            ho=self.num_heads, 
            v=self.head_dim * (self.order + 1)
```

amounts to:

```
b, d, seqlen = uc.shape
ho = self.num_blocks
v = self.head_dim * (self.order + 1)
z = self.num_blocks
l = seqlen // z
uc.reshape((b, ho, v, z, l)
```

so it reshapes the embedding dim into (heads, head size)